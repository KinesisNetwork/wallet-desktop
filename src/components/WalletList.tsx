import * as React from 'react'

import { RootRoutes, Wallet } from '@types'
import { NavLink } from 'react-router-dom'

export interface Props {
  wallets: Wallet[]
  activeWallet: Wallet | null
  selectWallet: (wallet: Wallet) => any
  deleteWallet: (wallet: Wallet) => any
  addWallet: () => any
}

class WalletList extends React.Component<Props> {
  constructor(props) {
    super(props)
  }

  renderWallets = () => {
    return this.props.wallets.map(wallet => {
      return (
        <WalletListItem
          key={wallet.publicKey}
          wallet={wallet}
          isActive={
            !!this.props.activeWallet && this.props.activeWallet.publicKey === wallet.publicKey
          }
          selectWallet={this.props.selectWallet}
          deleteWallet={this.props.deleteWallet}
        />
      )
    })
  }

  render() {
    return (
      <nav className="panel">
        <p className="panel-heading wallet-heading">Accounts</p>
        {this.renderWallets()}
        <div className="panel-block">
          <NavLink to={RootRoutes.create} className="button is-primary is-outlined is-fullwidth">
            Add Account
          </NavLink>
        </div>
      </nav>
    )
  }
}

interface WalletListItemProps {
  wallet: Wallet
  selectWallet: (wallet: Wallet) => any
  deleteWallet: (wallet: Wallet) => any
  isActive: boolean
}

const WalletListItem: React.SFC<WalletListItemProps> = ({
  wallet,
  selectWallet,
  deleteWallet,
  isActive,
}) => (
  <NavLink
    to={RootRoutes.dashboard}
    className={`panel-block`}
    activeClassName="is-active"
    onClick={() => selectWallet(wallet)}
    isActive={match => match && isActive}
  >
    <span className="panel-icon">
      <i className="fas fa-book" />
    </span>
    <span className="info" style={{ flexGrow: 2 }}>
      {wallet.accountName}
    </span>
    <button className="button is-small is-danger" onClick={() => deleteWallet(wallet)}>
      <span className="icon is-small">
        <i className="fas fa-trash-alt" />
      </span>
    </button>
  </NavLink>
)

export { WalletListItem, WalletList }
