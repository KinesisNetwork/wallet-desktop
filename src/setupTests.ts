import { configure } from 'enzyme'
import * as Adapter from 'enzyme-adapter-react-16'

export default configure({ adapter: new Adapter() })
