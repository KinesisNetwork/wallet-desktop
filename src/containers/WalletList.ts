import { connect } from 'react-redux'

import { changeWalletView, deleteWallet, selectWallet } from '@actions'
import { getPasswordConfirmation } from '@components/PasswordConfirmation'
import { WalletList as WalletPresentation } from '@components/WalletList'
import { Dispatch, RootState } from '@store'
import { RootRoutes, Wallet, WalletView } from '@types'

const mapStateToProps = ({ wallets, router }: RootState) => ({
  activeWallet: router.location.pathname.startsWith(RootRoutes.dashboard)
    ? wallets.activeWallet
    : null,
  wallets: wallets.walletList,
})

const mapDispatchToProps = (dispatch: Dispatch) => ({
  addWallet: () => dispatch(changeWalletView(WalletView.create)),
  selectWallet: (walletIndex: Wallet) => dispatch(selectWallet(walletIndex)),
  deleteWallet: async (wallet: Wallet) => {
    const isSureToDelete = await sweetAlert({
      buttons: ['Go back', 'Delete Account'],
      dangerMode: true,
      icon: 'warning',
      text: 'Once deleted, you will not be able to recover this account.',
      title: 'Are you sure?',
    })
    if (!isSureToDelete) {
      return
    }

    const { value: decryptedPrivateKeyOrEmpty } = await getPasswordConfirmation(wallet)
    if (decryptedPrivateKeyOrEmpty) {
      dispatch(deleteWallet({ ...wallet, decryptedPrivateKey: decryptedPrivateKeyOrEmpty }))
    }
  },
})

export const WalletList = connect(
  mapStateToProps,
  mapDispatchToProps,
)(WalletPresentation)
