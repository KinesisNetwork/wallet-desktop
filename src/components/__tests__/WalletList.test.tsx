import { shallow } from 'enzyme';
import * as React from 'react'

import { WalletList } from '@components/WalletList'
import '../../setupTests'

describe('WalletList', () => {
  let props

  beforeEach(() => {
    props = {
      activeWallet: {},
      addWallet: () => null,
      deleteWallet: () => null,
      selectWallet: () => null,
      wallets: [
        {
          name: 'Fred',
          publicKey: '123'
        },
        {
          name: 'Bob',
          publicKey: '456'
        }
      ]
    }
  })

  it('renders correctly', () => {
    const wrapper = shallow(<WalletList {...props} />)

    expect(wrapper).toMatchSnapshot()
  })
})
