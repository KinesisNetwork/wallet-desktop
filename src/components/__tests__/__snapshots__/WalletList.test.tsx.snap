// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletList renders correctly 1`] = `
<nav
  className="panel"
>
  <p
    className="panel-heading wallet-heading"
  >
    Accounts
  </p>
  <WalletListItem
    deleteWallet={[Function]}
    isActive={false}
    key="123"
    selectWallet={[Function]}
    wallet={
      Object {
        "name": "Fred",
        "publicKey": "123",
      }
    }
  />
  <WalletListItem
    deleteWallet={[Function]}
    isActive={false}
    key="456"
    selectWallet={[Function]}
    wallet={
      Object {
        "name": "Bob",
        "publicKey": "456",
      }
    }
  />
  <div
    className="panel-block"
  >
    <button
      className="button is-primary is-outlined is-fullwidth"
      onClick={[Function]}
    >
      Add Account
    </button>
  </div>
</nav>
`;
